scanner/decisions/0001-scannerv4-apis.md

```diff
@@ -0,0 +1,93 @@
+# 0001 - ScannerV4 APIs definition
+
+- **Status:** Accepted
+- **Created:** [2023-05-09 Tue]
+
+## Context
+
+We are transitioning Stackrox Scanner to use [ClairCore](https://github.com/quay/claircore) as its underlying scanning engine. The goal is to align vulnerability scanning results between Quay/Clair and Stackrox. The transition will be gradual, starting with container image scanning. During this period, the existing Scanner (ScannerV2) and the new ClairCore-based Scanner (ScannerV4) will be simultaneously deployed.
+
+The move to ClairCore and the adoption of ScannerV4 bring the need for new API definitions. Also, ClairCore allows for Clair deployment modes that better support the secured cluster use-case (local scanning), horizontal scalability, and high-availability scenarios. The new APIs ideally should align with ClairCore's capabilities while minimizing changes in Central.
+
+## Decision
+
+ScannerV4 APIs will exclusively use gRPC. ScannerV4 APIs are not backward compatible with ScannerV2.
+
+ScannerV4 will offer modes of operation [akin to Clair's deployment models](https://quay.github.io/clair/howto/deployment.html). Each mode will implement different gRPC services, reflecting the underlying separation between ClairCore's [libindex](https://pkg.go.dev/github.com/quay/claircore/libindex#Libindex) and [libvuln](https://pkg.go.dev/github.com/quay/claircore/libvuln#Libvuln). They will be named "indexer" and "matcher". Both modes can be enabled concurrently.
+
+ScannerV4 will use "index reports" and "vulnerability reports" data models, similar to ClairCore's. Both types will link to a "scannable resource" (e.g., a container image) using a "hash_id" managed by the clients. It uniquely identifies the resource's manifest.
+
+ScannerV4 will replace ScannerV2 as the container image scanner in Central. A new image integration for ScannerV4 will be created and replace ScannerV2's integration (type "clarify"), activated by the feature flag: `ROX_SCANNER_V4_ENABLED`. The "clarify" type will continue to be used for orchestrator and node scanning.
+
+ScannerV4 gRPC endpoints will use GRPC status codes to communicate the return status in case of non-successful responses.
+
+ScannerV4 will also provide endpoints for health and observability checks, which are not detailed in this ADR.
+
+### ScannerV4 API Endpoints
+
+The endpoints are versioned at `v4` to align with Clair:
+
+1. `scanner.v4.Indexer/CreateIndex`: Create a manifest of a specified resource and create an index or re-index. Idempotent while creation is being executed. Synchronous and tied to the client's request. Returns [`IndexReport`](https://github.com/quay/claircore/blob/v1.4.18/indexreport.go#L19)
+2. `scanner.v4.Indexer/GetIndex`: Retrieve or check index existence. Returns [`IndexReport`](https://github.com/quay/claircore/blob/v1.4.18/indexreport.go#L19).
```

Comment:
Let me know your stance on this in the pervious comment, and I will update here.

> I think it may make more sense to split this into two different RPCs. I'm just curious how it'd look if we were to turn this into an HTTP HEAD (test for existence) and GET (get the report)

The rationale was not to support HTTP. With the HTTP gateway in mind, I think we must have two separate endpoints.

Reply:
I feel like `check_only` seems a bit awkward for `GetIndexRequest`. I feel like having a separate `HasIndexReportRequest` would make more sense. Maybe it woul depend on what the response looks like. Speaking of which, I don't think I see any definitions for responses?

Reply:
> Speaking of which, I don't think I see any definitions for responses?

Let's iterate first before we specify the responses. For now, mentioning we will follow Clair's types is sufficient.

Reply:
> I feel like check_only seems a bit awkward for GetIndexRequest. I feel like having a separate HasIndexReportRequest would make more sense.

what are your thoughts on this part?

Reply:
I will onboard a new endpoint to open the door for HTTP-gateway.

Reply:
77252800a7..88e2a1c33a